src/server/engine_shard_set.cc

```diff
@@ -39,8 +49,260 @@ using absl::GetFlag;
+            << std::dec << OpTypeToString(params.object_type)
+            << " or was not allcoated on the heap";
+    return false;
+  }
+
+  params.slice->PreUpdate(index, it);
+  auto key = it->first.ToString();
+  auto state = ReallocateValueImpl(it->second);
+  params.slice->PostUpdate(index, it, key);
+  return state;
+}
+
+auto MemoryReallocate(PrimeIterator it, const MemoryDefragParams& params, DbIndex index)
+    -> MemDefragState {
+  if (!IsValid(it)) {
+    return MemDefragState::REQUIRED;
+  }
+  if (MatchType(it->second.ObjType(), params.object_type)) {
+    auto ptr = it->second.RObjPtr();
+    return ReallocateValue(it, ptr, params, index) ? MemDefragState::REQUIRED
+                                                   : MemDefragState::NOT_REQUIRED;
+  }
+  return MemDefragState::NOT_REQUIRED;
+}
+
+auto DefragMemory(const MemoryDefragParams& params, DbIndex index)
+    -> std::tuple<MemDefragState, uint64_t> {
+  constexpr size_t CURSOR_COUNT_RELATION = 50;
+
+  auto shard_id = params.slice->shard_id();
+  DCHECK(params.slice->IsDbValid(index));
+
+  unsigned cnt = 0;
+  std::size_t max = params.slice->DbSize(index) / CURSOR_COUNT_RELATION;
+  VLOG(1) << "PrimeTable shared id/index " << params.slice->shard_id() << "/" << index << " has "
+          << params.slice->DbSize(index) << " elements in it";
+
+  PrimeTable::Cursor cur = params.cursor;
+  auto [prime_table, expire_table] = params.slice->GetTables(index);
+  MemDefragState state = max > 0 ? MemDefragState::REQUIRED : MemDefragState::NOT_REQUIRED;
+
+  while (cnt < max && state != MemDefragState::NOT_REQUIRED) {
+    cur = prime_table->Traverse(
+        cur, [&](PrimeIterator it) { state = MemoryReallocate(it, params, index); });
+    if (!cur) {
+      state = MemDefragState::NOT_REQUIRED;
+    }
+  }
+
+  if (cnt > 0) {
+    VLOG(1) << shard_id << ": ran the memory re-ordering task for " << cnt << " times - cursor "
+            << cur.value();
+  }
+
+  return {state, cur.value()};
+}
+
+auto roundOff(double n) -> double {
+  double d = n * 100.0;
+  int i = d + 0.5;
+  d = (float)i / 100.0;
+  return d;
+}
+
+auto Mem2String(uint64_t mem) -> std::string {
+  static const char* SIZES[] = {"B", "KB", "MB", "GB"};
+  size_t div = 0;
+  size_t rem = 0;
+
+  while (mem >= 1024 && div < (sizeof SIZES / sizeof *SIZES)) {
+    rem = (mem % 1024);
+    div++;
+    mem /= 1024;
+  }
+
+  double size_d = (float)mem + (float)rem / 1024.0;
+  string result = std::to_string(roundOff(size_d)) + " " + SIZES[div];
+  return result;
+}
+
+inline auto MemUsageChanged(uint64_t current, uint64_t now) -> bool {
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+  // we only care whether the memory usage is going up
+  // otherwise we don't really case
+  return (current * commit_use_threshold) < float(now);
+}
+
+auto IsMemDefragRequired(uint16_t id, uint64_t* current_commit, uint64_t* current_use, int count)
```

Comment:
This function checks whether this we need to run the operation. We are also checking whether we have memory changes. In case we are doing the operation, we will save the current memory usage. Without this, it seems that we are getting into endless loop, as the defrag don't change memory usage.

Reply:
How about the following state machine:
```
bool should_traverse = we_traversed_last_time && cursor !=0;
if (!should_traverse)  {
  should_traverse = ShouldTriggerDefragByMemoryStats();
}

if (should_traverse) {
  cursor = Traverse(cursor);
}
```

if we still have an endless loop I would love to understand why given that we enable DECOMMIT option now.